### ISSUE TITLE
[Bug Report] "there are no relocation sections" error after upgrading tt-metal

### ISSUE THREAD

**Issue by marty1885** [tenstorrent/tt-metal#14619]
**Describe the bug**

Likely related to #12752. After upgrading tt-metal this morning, I get the following error when running my GGML backend.

```
Backend 1/2: METALIUM0                                                                                                                                 
  Device description: Tenstorrent Wormhole N300                                                                                                        
  Device memory: 12288 MB (12288 MB free)                                                                                                              
                                                                                                                                                       
  ABS(type=f32,ne_a=[128,2,2,2],v=0): terminate called after throwing an instance of 'std::runtime_error'                                              
  what():  TT_THROW @ /home/marty/Documents/tt/tt-metal/tt_metal/llrt/tt_elffile.cpp:711: tt::exception                                                info:                                                                                                                                                  
{}: there are no relocation sections                                                                                                                   
/home/marty/Documents/tt/tt-metal//built/6176/kernels/tilize/10935209067306454572//trisc0/trisc0.elf                                                   
backtrace:                                                                                                                                             
 --- ll_api::memory::memory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, ll_api::memory::Relocate)          
 --- tt::llrt::get_risc_binary(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, ll_api::memory::Pa
ckSpans, ll_api::memory::Relocate)                                                                                                                     
 --- tt::tt_metal::v0::ComputeKernel::read_binaries(tt::tt_metal::v0::Device*)                                                                         
 --- std::function<void ()>::operator()() const                                                                                                        
 --- void std::__invoke_impl<void, std::function<void ()>&>(std::__invoke_other, std::function<void ()>&)                                              
 --- std::__invoke_result<std::function<void ()>&>::type std::__invoke<std::function<void ()>&>(std::function<void ()>&)                               
 --- void std::_Bind<std::function<void ()> ()>::__call<void>(std::tuple<>&&, std::_Index_tuple<>)                                                     
 --- void std::_Bind<std::function<void ()> ()>::operator()<, void>()                                                                                  
 --- void std::__invoke_impl<void, std::_Bind<std::function<void ()> ()>&>(std::__invoke_other, std::_Bind<std::function<void ()> ()>&)
 --- std::enable_if<is_invocable_r_v<void, std::_Bind<std::function<void ()> ()>&>, void>::type std::__invoke_r<void, std::_Bind<std::function<void ()>
 ()>&>(std::_Bind<std::function<void ()> ()>&)
 --- std::__future_base::_Task_state<std::_Bind<std::function<void ()> ()>, std::allocator<int>, void ()>::_M_run()::{lambda()#1}::operator()() const
 --- std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_bas
e::_Task_state<std::_Bind<std::function<void ()> ()>, std::allocator<int>, void ()>::_M_run()::{lambda()#1}, void>::operator()() const
 --- std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter> std::__invoke_impl<std::unique_ptr<std::__future_ba
se::_Result<void>, std::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std
::__future_base::_Result_base::_Deleter>, std::__future_base::_Task_state<std::_Bind<std::function<void ()> ()>, std::allocator<int>, void ()>::_M_run(
)::{lambda()#1}, void>&>(std::__invoke_other, std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_
Result_base::_Deleter>, std::__future_base::_Task_state<std::_Bind<std::function<void ()> ()>, std::allocator<int>, void ()>::_M_run()::{lambda()#1}, v
oid>&)
```

**To Reproduce**

I don't have a particular way to replicate it as it seems to be dependent on which version of tt-metal I upgraded from. But in generally

1. Run any TTNN application with persistent cache enabled.
2. Update tt-metal and rebuilt without cleaning the persistent cache
1. Run any TTNN application with persistent cache enabled.
4. See the error

**Expected behavior**
TTNN should automatically invalidate the outdated object file.

**Screenshots**

**Please complete the following environment information:**
- OS: Arch Linux
- commit: a2c5123dd69f75a9075d8a4772a1337667f8bb48
- Cards: WH N300

**Additional context**

Deleting `$TT_METAL_HOME/built` solves the problem.

**Comment by pgkeller** [comment]
oops, assigned Nathan on a skim as it is related to the new elf file loader, but this looks like a stale cache issue

**Comment by pgkeller** [comment]
Here's the plan:
 - take the build id from libtt_metal.so and use this in the kernel path: "built/<id>/xyz"
 - at init remove all built kernels that don't match the build id w/ an env var that overrides this behavior
 - only find files in the cache if the build ids match

**Comment by pgkeller** [comment]
assigned @spoojaryTT to handle this after current task

**Comment by spoojaryTT** [comment]
Fix merged in 3d78f0c.
